src/corefx/System.Globalization.Native/localeNumberData.cpp

```diff
@@ -31,9 +47,259 @@ enum LocaleNumberData : int32_t
+
+	for (int i = 0; i < patternsCount; i++)
+	{
+		if (strcmp(charPattern, patterns[i]) == 0)
+		{
+			return i;
+		}
+	};
+
+	assert(false); // should have found a valid pattern
+	return -1;
+}
+
+/*
+Function:
+GetCurrencyNegativePattern
+
+Implementation of NumberFormatInfo.CurrencyNegativePattern.
+Returns the pattern index.
+*/
+int GetCurrencyNegativePattern(const Locale &locale)
+{
+	const int DEFAULT_VALUE = 0;
+	static const char* Patterns[] = { "(Cn)", "-Cn", "C-n", "Cn-", "(nC)", "-nC", "n-C", "nC-", "-n C", "-C n", "n C-", "C n-", "C -n", "n- C", "(C n)", "(n C)" };
+	UErrorCode status = U_ZERO_ERROR;
+
+	LocalPointer<NumberFormat> format(NumberFormat::createInstance(locale, UNUM_CURRENCY, status));
+	if (U_FAILURE(status))
+	{
+		assert(false);
+		return DEFAULT_VALUE;
+	}
+
+	int value = GetPattern(static_cast<DecimalFormat*>(format.getAlias()), Patterns, ARRAY_LENGTH(Patterns), true);
+	return (value >= 0) ? value : DEFAULT_VALUE;
+}
+
+/*
+Function:
+GetCurrencyPositivePattern
+
+Implementation of NumberFormatInfo.CurrencyPositivePattern.
+Returns the pattern index.
+*/
+int GetCurrencyPositivePattern(const Locale &locale)
+{
+	const int DEFAULT_VALUE = 0;
+	static const char* Patterns[] = { "Cn", "nC", "C n", "n C" };
+	UErrorCode status = U_ZERO_ERROR;
+
+	LocalPointer<NumberFormat> format(NumberFormat::createInstance(locale, UNUM_CURRENCY, status));
+	if (U_FAILURE(status))
+	{
+		assert(false);
+		return DEFAULT_VALUE;
+	}
+
+	int value = GetPattern(static_cast<DecimalFormat*>(format.getAlias()), Patterns, ARRAY_LENGTH(Patterns), false);
+	return (value >= 0) ? value : DEFAULT_VALUE;
+}
+
+/*
+Function:
+GetNumberNegativePattern
+
+Implementation of NumberFormatInfo.NumberNegativePattern.
+Returns the pattern index.
+*/
+int GetNumberNegativePattern(const Locale &locale)
+{
+	const int DEFAULT_VALUE = 1;
+	static const char* Patterns[] = { "(n)", "-n", "- n", "n-", "n -" };
+	UErrorCode status = U_ZERO_ERROR;
+
+	LocalPointer<NumberFormat> format(NumberFormat::createInstance(locale, UNUM_DECIMAL, status));
+	if (U_FAILURE(status))
+	{
+		assert(false);
+		return DEFAULT_VALUE;
+	}
+
+	int value = GetPattern(static_cast<DecimalFormat*>(format.getAlias()), Patterns, ARRAY_LENGTH(Patterns), true);
+	return (value >= 0) ? value : DEFAULT_VALUE;
+}
+
+/*
+Function:
+GetPercentNegativePattern
+
+Implementation of NumberFormatInfo.PercentNegativePattern.
+Returns the pattern index.
+*/
+int GetPercentNegativePattern(const Locale &locale)
+{
+	const int DEFAULT_VALUE = 0;
+	static const char* Patterns[] = { "-n %", "-n%", "-%n", "%-n", "%n-", "n-%", "n%-", "-% n", "n %-", "% n-", "% -n", "n- %" };
+	UErrorCode status = U_ZERO_ERROR;
+
+	LocalPointer<NumberFormat> format(NumberFormat::createInstance(locale, UNUM_PERCENT, status));
+	if (U_FAILURE(status))
+	{
+		assert(false);
+		return DEFAULT_VALUE;
+	}
+
+	int value = GetPattern(static_cast<DecimalFormat*>(format.getAlias()), Patterns, ARRAY_LENGTH(Patterns), true);
```

Comment:
Yes unless the design\code changes internally in ICU. The current code path always creates a DecimalFormat and at least in one location when consuming the same APIs I am, ICU creates a percent (or currency) and then does a C-style cast. Originally, I used a dynamic_cast but there appears to be a problem with C++ run-time-type-information as it was a hard crash when loading the .so file.